profiling/src/collections/string_table/mod.rs

```diff
@@ -0,0 +1,260 @@
+        // with endpoint + code hotspots, we'd have at least these strings:
+        // - ""
+        // - At least one sample type
+        // - At least one sample unit--already at 3 without any samples.
+        // - "local root span id"
+        // - "span id"
+        // - "trace endpoint"
+        // - A file and/or function name per frame.
+        // So with a capacity like 3, we end up reallocating a bunch on or
+        // before the very first sample. The number here is not fine-tuned,
+        // just skipping some obviously bad, tiny sizes.
+        strings.reserve(32);
+
+        // Always hold the empty string as item 0. Do not insert it via intern
+        // because that will try to allocate zero-bytes from the storage,
+        // which is sketchy.
+        strings.insert("");
+
+        Self { bytes, strings }
+    }
+
+    /// Returns the number of strings currently held in the string table.
+    #[inline]
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.strings.len()
+    }
+
+    /// Adds the string to the string table if it isn't present already, and
+    /// returns a [StringId] that corresponds to the order that this string
+    /// was originally inserted.
+    ///
+    /// # Panics
+    /// This panics if the allocator fails to allocate. This could happen for
+    /// a few reasons:
+    ///  - It failed to acquire a chunk.
+    ///  - The string is too large (larger than a chunk size, for example)
+    pub fn intern(&mut self, str: &str) -> StringId {
+        let set = &mut self.strings;
+        match set.get_index_of(str) {
+            Some(offset) => StringId::from_offset(offset),
+            None => {
+                // No match. Get the current size of the table, which
+                // corresponds to the StringId it will have when inserted.
+                let string_id = StringId::from_offset(set.len());
+
+                // Allocate the string with alignment of 1 so that bytes are
+                // not wasted between strings.
+                let uninit_ptr = {
+                    // SAFETY: this layout matches an existing string, it must
+                    // be a valid layout since it already exists.
+                    let layout =
+                        unsafe { Layout::from_size_align(str.len(), 1).unwrap_unchecked() };
+                    self.bytes.allocate(layout).unwrap()
+                };
+
+                // Copy the bytes of the string into the allocated memory.
+                // SAFETY: this is guaranteed to not be overlapping because an
+                // allocator must not return aliasing bytes in its allocations.
+                unsafe {
+                    let src = str.as_ptr();
+                    let dst = uninit_ptr.as_ptr() as *mut u8;
+                    let count = str.len();
+                    core::ptr::copy_nonoverlapping(src, dst, count)
```

Comment:
Minor: More of a structure kinda of thing, I'd kinda expect these two things (especially the unsafe bits) to be internal to `ChainAllocator`, rather than having the users need to get their hands dirty.

Reply:
In Rust, allocators don't initialize things. They deal with the shapes of the things (`Layout`) and give you bytes. The caller is responsible to initialize them.

I could create a custom `String` type that has a method like `String::new_in(str: &str, allocator: &impl Allocator)`. I don't feel like it's worth it for _this_ purpose only, but if there's some other reason you can think of to want it, then we could do that. We could bind the lifetime of the custom String struct, but then we'd have to deal with self-referential structs, which are also unsafe. The [ouroboros](https://crates.io/crates/ouroboros) crate can help with that, but it's a bit complex. We can do it if you or others think it's worth it.

Reply:
I did a little refactoring to pull the alloc + initialization into a single operation so the interning code reads in a better way. Let me know what you think. We can make these `'self -> 'static -> 'self` operations safe, if we really want to. It's just a lot of boilerplate to use ouroboros for it.

Reply:
I think the new version is clearer, but I'm still kinda curious (and I don't think any of these are blockers, it's fine as-is):

* Currently the allocation exposes its guts. Then the caller uses them for what it wants. I was kinda wondering if it would make sense to do it the other way -- by telling the allocator: here's a pointer, and here's length, return me a pointer that's a result of copying the pointer/length inside yourself. Maybe even use generics to be able to return a pointer to whatever was passed in, rather than a generic pointer?

* The transmute in `intern` is it just a really fancy way of saying "never try to deallocate this string directly, as I'm managing the memory in another way"?

Reply:
- The allocation API is a mirror of what is in the work-in-progress Allocator API for standard Rust. It's not something I think we should change.
- The transmute of the lifetime doesn't deal with the deallocation in this case, because it's just a reference either way. The transmute is to tell the compiler that the lifetime of these references outlives the data it refers to. As mentioned, we can safely do this with self-referential lifetimes but Rust doesn't have those.